contrib/lib/src/databases.rs

```diff
@@ -780,6 +810,184 @@ impl Poolable for memcache::Client {
+    _marker: PhantomData<fn() -> K>,
+}
+
+/// Unstable internal details of generated code for the #[database] attribute.
+///
+/// This type is implemented here instead of in generated code to ensure all
+/// types are properly checked.
+#[doc(hidden)]
+pub struct Connection<K, C: Poolable> {
+    pool: ConnectionPool<K, C>,
+    connection: Option<r2d2::PooledConnection<C::Manager>>,
+    _permit: Option<OwnedSemaphorePermit>,
+    _marker: PhantomData<fn() -> K>,
+}
+
+// A wrapper around spawn_blocking that propagates panics to the calling code
+async fn run_blocking<F, R>(job: F) -> R
+where
+    F: FnOnce() -> R + Send + 'static,
+    R: Send + 'static,
+{
+    match tokio::task::spawn_blocking(job).await {
+        Ok(ret) => ret,
+        Err(e) => match e.try_into_panic() {
+            Ok(panic) => std::panic::resume_unwind(panic),
+            Err(_) => unreachable!("spawn_blocking tasks are never canceled"),
+        }
+    }
+}
+
+impl<K: 'static, C: Poolable> ConnectionPool<K, C> {
+    #[inline]
+    pub fn fairing(fairing_name: &'static str, config_name: &'static str) -> impl Fairing {
+        AdHoc::on_attach(fairing_name, move |mut rocket| async move {
+            let config = database_config(config_name, rocket.config().await);
+            let pool = config.map(|c| (c.pool_size, C::pool(c)));
+
+            match pool {
+                Ok((size, Ok(pool))) => {
+                    let managed = ConnectionPool::<K, C> {
+                        pool,
+                        semaphore: Arc::new(Semaphore::new(size as usize)),
+                        _marker: PhantomData,
+                    };
+                    Ok(rocket.manage(managed))
+                },
+                Err(config_error) => {
+                    rocket::logger::error(
+                        &format!("Database configuration failure: '{}'", config_name));
+                    rocket::logger::error_(&config_error.to_string());
+                    Err(rocket)
+                },
+                Ok((_, Err(pool_error))) => {
+                    rocket::logger::error(
+                        &format!("Failed to initialize pool for '{}'", config_name));
+                    rocket::logger::error_(&format!("{:?}", pool_error));
+                    Err(rocket)
+                },
+            }
+        })
+    }
+
+    async fn get(&self) -> Result<Connection<K, C>, ()> {
+        // TODO: timeout duration
+        let permit = match tokio::time::timeout(
+            self.pool.connection_timeout(),
+            self.semaphore.clone().acquire_owned()
+        ).await {
+            Ok(p) => p,
+            Err(_) => {
+                error_!("Failed to get a database connection within the timeout.");
+                return Err(());
+            }
+        };
+
+        let pool = self.pool.clone();
+
+        // TODO: timeout duration
+        match run_blocking(move || pool.get_timeout(std::time::Duration::from_secs(0))).await {
```

Comment:
We have two opportunities to set a timeout here:
* When acquiring the semaphore. This waits on other connection handles to be released.
* When retrieving a connection from the pool. This should usually complete immediately, but it could need to wait on the network if connections are being / need to be replenished.

Previously there was only one timeout, that waited on other connections to be released. I tried to do the most straightforward conversion, but should we relax the timeout on checking out a connection to a few seconds?

Reply:
> We have two opportunities to set a timeout here:
> 
> * When acquiring the semaphore. This waits on other connection handles to be released.
> * When retrieving a connection from the pool. This should usually complete immediately, but it could need to wait on the network if connections are being / need to be replenished.
> 
> Previously there was only one timeout, that waited on other connections to be released. I tried to do the most straightforward conversion, but should we relax the timeout on checking out a connection to a few seconds?

Ideally the latter would be configurable, with a default of `0`, and should be set on the `r2d2::Builder` so that `get()` automatically uses it. The config revamp will make doing this much easier. I propose we make a note of this and wait until I can get those changes in.